pdr_backend/lake/etl.py

```diff
@@ -56,27 +77,62 @@ def _drop_temp_sql_tables(self):
         db = DuckDBDataStore(self.ppss.lake_ss.lake_dir)
         for dataclass in _ETL_REGISTERED_LAKE_TABLES:
             table_name = dataclass.get_lake_table_name()
-            db.drop_view(ETLTable(table_name).fullname)
-            db.drop_table(TempTable(table_name).fullname)
+            db.drop_table(TempTable(table_name).table_name)
+            db.drop_table(NewEventsTable(table_name).table_name)
+            db.drop_table(UpdateEventsTable(table_name).table_name)
+            db.drop_table(TempUpdateTable(table_name).table_name)
 
-    def _move_from_temp_tables_to_live(self):
+    def _do_bronze_swap_to_prod_atomic(self):
         """
         @description
-            Move the records from our bronze temporary tables to live, in-production tables
+            Merge the bronze ETL tables against prod database
+            This needs to become a single transaction per table
+            Such that it remains atomic, and is able to resume if it fails
         """
-
         db = DuckDBDataStore(self.ppss.lake_ss.lake_dir)
-        for dataclass in _ETL_REGISTERED_LAKE_TABLES:
-            table_name = dataclass.get_lake_table_name()
-            logger.info("move table %s to live", table_name)
-            temp_table = TempTable(table_name)
-            permanent_table = NamedTable(table_name)
 
-            if db.table_exists(temp_table.fullname):
-                db.move_table_data(temp_table, permanent_table)
+        bronze_tables = [BronzePrediction]
+        for table in bronze_tables:
+            prod_table = Table.from_dataclass(table)
+            new_events_table = NewEventsTable.from_dataclass(table)
+            update_events_table = UpdateEventsTable.from_dataclass(table)
+            temp_update_table = TempUpdateTable.from_dataclass(table)
+
+            if db.table_exists(update_events_table.table_name):
+                # Insert new records into live tables
+                # We don't know if the table exists or not, so get the query string
+                temp_to_prod_query = db.get_query_move_table_data(
+                    new_events_table, prod_table
+                )
 
-                db.drop_view(ETLTable(table_name).fullname)
-                db.drop_table(temp_table.fullname)
+                # We'll also get the query string from the helper
+                drop_records_from_table_by_id_query = (
+                    db.get_query_drop_records_from_table_by_id(
+                        prod_table.table_name, temp_update_table.table_name
+                    )
+                )
+
+                # Here, the table needs to exist. So we'll build our own insert statement.
+                temp_update_to_prod_query = f"""
+                INSERT INTO {prod_table.table_name} SELECT * FROM {temp_update_table.table_name};
+                """
+
+                # We then need to drop the rows after all the ETL is complete
+                cleanup_query = f"""
```

Comment:
I think these assert/post-work queries can be separated to a different function to simplify reading and understanding.

Reply:
I disagree.

This is part of the whole swap strategy/algorithm. It breaks LoB "locality of behavior" and makes very specific functionality/logic look generic.